README.md
> This library is under active development. Please [email us](mailto:shaun@stripe.com) if you’re ready to use it in a production environment.

# Mobile Viewport Control

This JavaScript library attempts to solve the hard problem of creating a
__[full-window modal]__ experience for mobile browsers.  The goal is to allow
it to be dropped into any page regardless of its viewport settings.  It does
this by attempting to control the mobile viewport and by hiding all elements
except the one you are trying to modalize.

[full-window modal]:http://uxmovement.com/mobile/why-you-should-avoid-using-modal-windows-on-mobile/

--

 <table>
  <tr>
    <td>[![scroll][scroll-gif]][scroll-gfy]</td>
    <td>[![zoom][zoom-gif]][zoom-gfy]</td>
  </tr>
  <tr>
    <td>__Figure 1.__ Scrolling</td>
    <td>__Figure 2.__ Zooming + Scrolling</td>
  </tr>
</table>

--

__Background.__ The _viewport_ is the visible part of the webpage on your
screen (see quirksmode [[1]] [[2]]).  It is controlled by two things: _scroll_
and _zoom_.  The user can swipe to scroll and pinch to zoom. A JavaScript
developer can control the scroll programmatically, but there is no officially
supported method for controlling the _zoom_ programmatically.  However, we do
have a de facto standard for limiting and initializing the zoom:

[1]:http://www.quirksmode.org/mobile/viewports2.html
[2]:http://www.quirksmode.org/mobile/metaviewport/

```html
<head>
  <meta name="viewport" content="initial-scale=1">
  <meta name="viewport" content="minimum-scale=0">
  <meta name="viewport" content="maximum-scale=10">
</head>
```

__Workaround.__ There is an undocumented feature that allows us to modify the
viewport state at runtime by adding and modifying viewport tags.  There are
caveats depending on the browser.  The goal of this library is to identify and
workaround these caveats until there is a standard for programmatically
controlling zoom.  We have researched and tested on multiple platforms, though
we are currently focusing support for Mobile Safari and WebViews on iOS 7,8,9.
(See __[Compatibility Testing](test/)__)

__Modals are the Use-Case.__ Viewport settings vary widely across different
webpages, making it hard to create drop-in modal experiences that work
everywhere (see [Stripe Checkout] and [Auth0 Lock]).  For browsers that cannot
create pop-up tabs and for pages where redirection is not an option, this
library can help fill in the gaps.

[Stripe Checkout]:https://stripe.com/checkout
[Auth0 Lock]:https://auth0.com/lock

--

 <table>
  <tr>
    <td>[![freeze][freeze-gif]][freeze-gfy]</td>
    <td>[![isolate][isolate-gif]][isolate-gfy]</td>
  </tr>
  <tr>
    <td>__Figure 3.__ Freezing and Thawing</td>
    <td>__Figure 4.__ Isolating the Modal</td>
  </tr>
 </table>

--

__The Abstraction.__ This library provides a limited abstraction for
controlling zoom: _freezing_ and _thawing_.  (See Figure 3). Freezing will lock
the zoom to a given scale.  Thawing will restore the viewport to what it was
before freezing, while also restoring the appropriate zoom limits (i.e. min and
max).  Thawing in this way is useful in the context of a modal.

__Isolating the Modal.__ The freeze/thaw functions are designed to be used in
conjunction with a third operation: _isolation_. (See Figure 4).  This is done
by passing an optional element ID to the freeze function, which you would
presumably wish to be the full-window modal.  This temporarily applies CSS
rules to ensure with high confidence that all elements except the given element
are collapsed (via `display:none`).  You can then style your element to fill
the window as a modal. Thawing will restore the page back to what it was
before.

## Usage

```
npm install mobile-viewport-control
```

```js
const viewport = require('mobile-viewport-control');
// ... or use `window.mobileViewportControl` if not used with package manager


// Freeze the viewport at a desired scale.
viewport.freeze(1.0, () => console.log('notified when frozen!'));

// Restore the viewport to what it was before freezing.
viewport.thaw(() => console.log('notified when thawed!'));
```

Additionally, to freeze the scroll area to a given element, you can pass an
element ID as the second argument.  We do this by temporarily hiding everything
else on the page.  Just make sure your element is a direct child of
`document.body`.

```js
viewport.freeze(1.0, 'myElementID');
```

## Testing

Please see __[Compatibility Testing](test/)__.

## License

[ISC License](LICENSE)

--

_Source code for the animated figures above: [mobile-viewport-control-viz](https://github.com/shaunstripe/mobile-viewport-control-viz)_


[scroll-gif]:https://zippy.gfycat.com/EnchantedPowerfulIndri.gif
[scroll-gfy]:https://gfycat.com/EnchantedPowerfulIndri
[zoom-gif]:https://zippy.gfycat.com/InconsequentialImperturbableBuckeyebutterfly.gif
[zoom-gfy]:https://gfycat.com/InconsequentialImperturbableBuckeyebutterfly
[freeze-gif]:https://fat.gfycat.com/YearlyBitesizedBangeltiger.gif
[freeze-gfy]:https://gfycat.com/YearlyBitesizedBangeltiger
[isolate-gif]:https://fat.gfycat.com/MetallicFrighteningApisdorsatalaboriosa.gif
[isolate-gfy]:https://gfycat.com/MetallicFrighteningApisdorsatalaboriosa


CHANGES.md
# Changes

## 0.3.1

- android webview support

## 0.3.0

- iOS - zoom bounds restored on thaw
- UIWebView - partial freeze support (double-tap or pinch-zoom auto-zooms)
- Android Firefox - scroll fixed on thaw
- Android Firefox - page width fixed on freeze
- Android Chrome/WebView - scroll/scale fixed on thaw

## 0.2.0

- add option for isolating an element when freezing

## 0.1.1

- restore specified viewport width on thaw
- export `getInitialViewport`

## 0.1.0

- initial release


test.js
var globalTest = {
  success: true,
  cases: [],
  waitLength: 100,
};

function assert(test) {
  test.success = test.actual === test.expected;

  globalTest.success &= test.success;
  globalTest.cases.push(test);

  document.body.style.background = globalTest.success ? "#0F0" : "#F00";

  if (test.success) {
    resultDiv.innerHTML += (
      "<div>"+
      "&#x2714; "+test.name+" is "+test.expected+
      "</div>"
    );
  }
  else {
    resultDiv.innerHTML += (
      "<div>"+
      "&#x2718; "+test.name+"<br>"+
      "<b>expected</b>: "+test.expected+"<br>"+
      "<b>actual</b>: "+test.actual+
      "</div>"
    );
  }
}


test/test.js
var globalTest = {
  success: true,
  totalPass: 0,
  cases: [],
  waitLength: 100,
  epsilon: 0.002, // scale equality threshold
};

if (window.runBtn) {
  runBtn.onclick = function() {
    runBtn.parentElement.removeChild(runBtn);
    globalTest.run();
  }
} else {
  setTimeout(function() {
    globalTest.run();
  }, 1);
}

function assert(test) {
  var isExact = test.actual === test.expected;
  var isClose = Math.abs(test.actual - test.expected) < globalTest.epsilon;

  test.success = isClose;
  if (test.success) {
    globalTest.totalPass++;
  }

  globalTest.success &= test.success;
  globalTest.cases.push(test);

  document.body.style.background = globalTest.success ? "#0F0" : "#F00";

  if (test.success) {
    resultDiv.innerHTML += (
      "<div>"+
      "&#x2714; "+test.name+" is "+test.expected+
      (!isExact && isClose ? "~" : "")+
      "</div>"
    );
  }
  else {
    resultDiv.innerHTML += (
      "<div>"+
      "&#x2718; "+test.name+"<br>"+
      "<b>expected</b>: "+test.expected+"<br>"+
      "<b>actual</b>: "+test.actual+
      "</div>"
    );
  }
}

function log(msg) {
  resultDiv.innerHTML += (
    "<div>"+
    "&#9758; "+msg+
    "</div>"
  );
}

function finishTesting() {
  resultDiv.innerHTML += (
    (globalTest.success ? "SUCCESS" : "FAILED") +
    " (" + globalTest.totalPass + "/" + globalTest.cases.length + " passed)"
  );
}

function testSequence(funcs) {
  if (funcs.length === 0) {
    finishTesting();
    return;
  }
  setTimeout(function(){
    funcs[0]();
    testSequence(funcs.slice(1));
  }, globalTest.waitLength);
}


proxy/test.js
(function(){
  var viewport = window.mobileViewportControl;
  if (viewport) {
    init();
  } else {
    require(['mobile-viewport-control'], function(module) {
      viewport = module;
      init();
    });
  }

  function init() {
    alert('Scroll and zoom for a bit. ' +
          '5 seconds after closing this, the viewport will freeze. ' +
          'Currently at:\n' +
          viewport.getPrettyInitialViewport());

    setTimeout(test, 5000);
  }

  function test() {
    var isolated = createIsolatedElement();
    var initial = viewport.getPrettyInitialViewport();
    viewport.freeze(1, isolated.id);
  }

  function createIsolatedElement() {
    var isolated = document.createElement('div');
    isolated.style.cssText = 'color: #000; text-align: left; font-size: 12px; padding: 20px;';
    isolated.id = 'myIsolatedElement';
    isolated.innerHTML = 'I should be the only visible thing, and you should not be able to pinch zoom or scroll.<br>';

    var btn = document.createElement('button');
    btn.innerHTML = 'Click me to restore!';
    btn.onclick = function() {
      document.body.removeChild(isolated);
      viewport.thaw();
    }
    isolated.appendChild(btn);

    document.body.appendChild(isolated);
    return isolated;
  }

  //alert('hello world:\n'+viewport.getInitialViewport());
}());


index.js
//
// Mobile Viewport Control v0.3.1
//
// Copyright (c) 2016 Shaun Williams
// ISC License
//
// GitHub: https://github.com/shaunstripe/mobile-viewport-control
//

//---------------------------------------------------------------------------
// JS Module Boilerplate
//---------------------------------------------------------------------------

(function (root, factory) {
  if (typeof define === 'function' && define.amd) {
    define('mobile-viewport-control', [], factory);
  }
  else if (typeof module === 'object' && module.exports) {
    module.exports = factory();
  }
  else {
    root.mobileViewportControl = factory();
  }
}(this, function() {

//---------------------------------------------------------------------------
// Getting/Setting Scroll position
//---------------------------------------------------------------------------

function getScroll() {
  return {
    top: window.pageYOffset || document.documentElement.scrollTop,
    left: window.pageXOffset || document.documentElement.scrollLeft
  };
}

function setScroll(scroll) {
  if (window.scrollTo) {
    window.scrollTo(scroll.left, scroll.top);
  } else {
    document.documentElement.scrollTop = scroll.top;
    document.documentElement.scrollLeft = scroll.left;
    document.body.scrollTop = scroll.top;
    document.body.scrollLeft = scroll.left;
  }
}

//---------------------------------------------------------------------------
// Getting Initial Viewport from <meta name='viewport'> tags
// but we also include implicit defaults.
//---------------------------------------------------------------------------

function getInitialViewport(withDefaults) {
  var viewport = {};

  if (withDefaults) {
    // These seem to be the defaults
    viewport = {
      'user-scalable': 'yes',
      'minimum-scale': '0',
      'maximum-scale': '10'
    };
  }

  var tags = document.querySelectorAll('meta[name=viewport]');
  var i,j,tag,content,keyvals,keyval;
  for (i=0; i<tags.length; i++) {
    tag = tags[i];
    content = tag.getAttribute('content');
    if (tag.id !== hookID && content) {
      keyvals = content.split(',');
      for (j=0; j<keyvals.length; j++) {
        keyval = keyvals[j].split('=');
        if (keyval.length === 2) {
          viewport[keyval[0].trim()] = keyval[1].trim();
        }
      }
    }
  }
  return viewport;
}

function getPrettyInitialViewport() {
  var initial = getInitialViewport();
  var keyvals = [];
  for (var prop in initial) {
    if (initial.hasOwnProperty(prop)) {
      keyvals.push({key:prop, val:initial[prop]});
    }
  }
  return (
    keyvals.sort(function(a,b) {
      if (a.key < b.key) return -1;
      if (a.key > b.key) return 1;
      return 0;
    }).map(function(kv) {
      return kv.key + '=' + kv.val;
    }).join(',\n')
  );
}

//---------------------------------------------------------------------------
// Calculating current viewport scale
// simplified from: http://menacingcloud.com/?c=viewportScale
//---------------------------------------------------------------------------

function getOrientation() {
  var degrees = window.orientation;
  var w = document.documentElement.clientWidth;
  var h = document.documentElement.clientHeight;
  if(degrees === undefined) {
    return (w > h) ? 'landscape' : 'portrait';
  }
  return (degrees % 180 === 0) ? 'portrait' : 'landscape';
}

function getOrientedScreenWidth() {
  var orientation = getOrientation();
  var sw = screen.width;
  var sh = screen.height;
  return (orientation === 'portrait') ? Math.min(sw,sh) : Math.max(sw,sh);
}

function getScale() {
  var visualViewportWidth = window.innerWidth;
  var screenWidth = getOrientedScreenWidth();
  return screenWidth / visualViewportWidth;
}


//---------------------------------------------------------------------------
// Get mobile OS
// from: http://stackoverflow.com/a/21742107
//---------------------------------------------------------------------------

function getMobileOS() {
  var userAgent = navigator.userAgent || navigator.vendor || window.opera;
  if (userAgent.match(/iPad/i) ||
      userAgent.match(/iPhone/i) ||
      userAgent.match(/iPod/i)) {
    return 'iOS';
  }
  else if (userAgent.match(/Android/i)) {
    return 'Android';
  }
}

function isFirefox() {
  var userAgent = navigator.userAgent || navigator.vendor || window.opera;
  return userAgent.match(/Firefox/i) ? true : false;
}

//---------------------------------------------------------------------------
// State and Constants
//---------------------------------------------------------------------------

// A unique ID of the meta-viewport tag we must create
// to hook into and control the viewport.
var hookID = '__mobileViewportControl_hook__';

// A unique ID of the CSS style tag we must create to
// add rules for hiding the body.
var styleID = '__mobileViewPortControl_style__';

// An empirical guess for the maximum time that we have to
// wait before we are confident a viewport change has registered.
var refreshDelay = 200;

// Original viewport state before freezing.
var originalScale;
var originalScroll;

// Classes we use to make our css selector specific enough
// to hopefully override all other selectors.
// (mvc__ = mobileViewportControl prefix for uniqueness)
var hiddenClasses = [
  'mvc__a',
  'mvc__lot',
  'mvc__of',
  'mvc__classes',
  'mvc__to',
  'mvc__increase',
  'mvc__the',
  'mvc__odds',
  'mvc__of',
  'mvc__winning',
  'mvc__specificity'
];

//---------------------------------------------------------------------------
// Isolating an Element
//---------------------------------------------------------------------------

function isolatedStyle(elementID) {
  var classes = hiddenClasses.join('.');
  return [
    // We effectively clear the <html> and <body> background
    // and sizing attributes.
    'html.' + classes + ',',
    'html.' + classes + ' > body {',
    '  background: #fff;',
    '  width: auto;',
    '  min-width: inherit;',
    '  max-width: inherit;',
    '  height: auto;',
    '  min-height: inherit;',
    '  max-height: inherit;',
    '  margin: 0;',
    '  padding: 0;',
    '  border: 0;',
    '}',
    // hide everything in the body...
    'html.' + classes + ' > body > * {',
    '  display: none !important;',
    '}',
    // ...except the given element ID
    'html.' + classes + ' > body > #' + elementID + ' {',
    '  display: block !important;',
    '}'
  ].join('\n');
}

function isolate(elementID) {
  // add classes to body tag to isolate all other elements
  var classes = hiddenClasses.join(' ');
  var html = document.documentElement;
  html.className += ' ' + classes;

  // add isolating style rules
  var style = document.createElement('style');
  style.id = styleID;
  style.type = 'text/css';
  style.appendChild(document.createTextNode(isolatedStyle(elementID)));
  document.head.appendChild(style);
}

function undoIsolate() {
  // remove isolating classes from body tag
  var classes = hiddenClasses.join(' ');
  var html = document.documentElement;
  html.className = html.className.replace(classes, '');

  // remove isolating style rules
  var style = document.getElementById(styleID);
  document.head.removeChild(style);
}

//---------------------------------------------------------------------------
// Freezing
//---------------------------------------------------------------------------

// Freeze the viewport to a given scale.
function freeze(scale) {
  // optional arguments
  var isolateID, onDone;

  // get optional arguments using their type
  var args = Array.prototype.slice.call(arguments, 1);
  if (typeof args[0] === 'string') {
    isolateID = args[0];
    args.splice(0, 1);
  }
  if (typeof args[0] === 'function') {
    onDone = args[0];
  }

  // save original viewport state
  originalScroll = getScroll();
  originalScale = getScale();

  // isolate element if needed
  if (isolateID) {
    isolate(isolateID);
    setScroll({x:0,y:0});
  }

  // validate scale
  // (we cannot freeze scale at 1.0 on Android)
  if (scale === 1) {
    scale = 1.002;
  }

  // add our new meta viewport tag
  var hook = document.getElementById(hookID);
  if (!hook) {
    hook = document.createElement('meta');
    hook.id = hookID;
    hook.name = 'viewport';
    document.head.appendChild(hook);
  }

  // When freezing the viewport, we still enable
  // user-scalability and allow a tight zooming
  // margin.  Without this, UIWebView would simply
  // ignore attempts to set the scale.  But with this
  // solution, the next time the user pinch-zooms
  // in this state, the viewport will auto-snap
  // to our scale.

  var includeWidth = (getMobileOS() === 'Android' && isFirefox());
  hook.setAttribute('content', [
    'user-scalable=yes',
    'initial-scale='+scale,
    'minimum-scale='+scale,
    'maximum-scale='+(scale+0.004),
    (includeWidth ? 'width=device-width' : null)
  ].filter(Boolean).join(','));

  if (onDone) {
    setTimeout(onDone, refreshDelay);
  }
}

//---------------------------------------------------------------------------
// Thawing
//---------------------------------------------------------------------------

function thawWebkit(hook, initial, onDone) {
  // Restore the user's manual zoom.
  hook.setAttribute('content', [
    'initial-scale='+originalScale,
    'minimum-scale='+originalScale,
    'maximum-scale='+originalScale
  ].join(','));

  // Restore the page's zoom bounds.
  hook.setAttribute('content', [
    'user-scalable='+initial['user-scalable'],
    'minimum-scale='+initial['minimum-scale'],
    'maximum-scale='+initial['maximum-scale'],
    (initial.width ? 'width='+initial.width : null)
  ].filter(Boolean).join(','));

  // Remove our meta viewport hook.
  document.head.removeChild(hook);

  setScroll(originalScroll);

  setTimeout(function() {
    if (onDone)
      onDone();
  }, refreshDelay);
}

function thawGecko(hook, initial, onDone) {
  // Restore the user's manual zoom.
  hook.setAttribute('content', [
    'initial-scale='+originalScale,
    'minimum-scale='+originalScale,
    'maximum-scale='+originalScale
  ].join(','));

  // Updating the scroll here is too early,
  // but it's used to force a refresh of the viewport
  // with our current desired scale.
  setScroll(originalScroll);

  setTimeout(function(){
    // Restore the page's zoom bounds.
    hook.setAttribute('content', [
      'user-scalable='+initial['user-scalable'],
      'minimum-scale='+initial['minimum-scale'],
      'maximum-scale='+initial['maximum-scale'],
      (initial.width ? 'width='+initial.width : null)
    ].filter(Boolean).join(','));

    // Restore the scroll again now that the scale is correct.
    setScroll(originalScroll);

    // Remove our meta viewport hook.
    document.head.removeChild(hook);

    if (onDone)
      onDone();
  }, refreshDelay);
}

function thawBlink(hook, initial, onDone) {
  hook.setAttribute('content', [
    'user-scalable='+initial['user-scalable'],
    // WebView does not support this:
    //'initial-scale='+originalScale
    'initial-scale='+initial['initial-scale'],
    'minimum-scale='+initial['minimum-scale'],
    'maximum-scale='+initial['maximum-scale'],
    (initial.width ? 'width='+initial.width : null)
  ].filter(Boolean).join(','));

  setScroll(originalScroll);

  setTimeout(function(){
    document.head.removeChild(hook);
    if (onDone)
      onDone();
  }, refreshDelay);
}

// Thaw the viewport, restoring the scale and scroll to what it
// was before freezing.
function thaw(onDone) {
  // restore body visibility
  var style = document.getElementById(styleID);
  if (style) {
    undoIsolate();
  }

  // exit if there is nothing to thaw
  var hook = document.getElementById(hookID);
  if (!hook) {
    return;
  }

  var initial = getInitialViewport(true);

  // thaw function defaults to webkit
  var thawFunc = thawWebkit;
  var os = getMobileOS();
  if (os === 'Android')  { thawFunc = isFirefox() ? thawGecko : thawBlink; }
  else if (os === 'iOS') { thawFunc = thawWebkit; }

  // call appropriate thaw function
  thawFunc(hook, initial, onDone);
}

//---------------------------------------------------------------------------
// Public API
//---------------------------------------------------------------------------

return {
  getInitialViewport: getInitialViewport,
  getPrettyInitialViewport: getPrettyInitialViewport,
  getScale: getScale,
  isolate: isolate,
  undoIsolate: undoIsolate,

  // stable
  version: '0.3.1',
  freeze: freeze,
  thaw: thaw
};

})); // end module scope


proxy/proxy.js
'use strict';

var port = 8081;

var fs = require('fs');
var colors = require('colors');
var Proxy = require('http-mitm-proxy');
var proxy = Proxy();

// get JS that we will inject into the webpages
var libJS = fs.readFileSync('../index.js','utf8');
var testJS = fs.readFileSync('test.js','utf8');

// get and print domain whitelist
var domains;

function getDomainWhiteList() {
  try {
    domains = fs.readFileSync('domains', 'utf8')
                .split('\n')
                .map(function(x) { return x ? x.trim() : x; })
                .filter(Boolean);
    if (domains.length === 0) {
      throw "domains file empty";
    }
    console.log('\nDomain Whitelist:')
    for (var i=0; i<domains.length; i++) {
      console.log(i, domains[i]);
    }
    console.log();
  }
  catch (e) {
    console.log(colors.red('\nERROR: ')+'Please create a whitelist "domains" file with at least one domain, one per line.');
    console.log('The listed domains will be the only ones we inject JS into.');
    process.exit(1);
  }
}

function isDomainWhiteListed(host) {
  var i;
  for (i=0; i<domains.length; i++) {
    if (host.indexOf(domains[i]) !== -1) {
      return true;
    }
  }
  return false;
}

function log(msg) {
  console.info(colors.cyan('INJECT_LOG: ')+msg);
}

function injectJS(host, url, body) {
  log(host+url);

  if (!isDomainWhiteListed(host)) {
    log('Domain not in whitelist. Ignoring.');
    return;
  }
  log('Domain in whitelist. Proceeding...');

  var bodyStr = body.toString();
  var tag = '</body>';

  log('Checking for '+tag+' tag...');
  if (bodyStr.indexOf(tag) === -1) {
    log('Could not inject JS because '+tag+' was missing');
    return;
  }

  log('Found tag. Injecting JS...');
  return bodyStr.replace(
    tag,
    '<script>' + libJS + '</script>' +
    '<script>' + testJS + '</script>' + tag
  );
}

proxy.onError(function(ctx, err, errorKind) {
  // ctx may be null
  var url = (ctx && ctx.clientToProxyRequest) ? ctx.clientToProxyRequest.url : '';
  console.error(errorKind + ' on ' + url + ':', err);
});

proxy.use(Proxy.gunzip);

proxy.onRequest(function(ctx, callback) {
  var host = ctx.clientToProxyRequest.headers.host;
  var url = ctx.clientToProxyRequest.url;

  var chunks = [];
  ctx.onResponseData(function(ctx, chunk, callback) {
    chunks.push(chunk);
    return callback(null, null); // don't write chunks to client response
  });
  ctx.onResponseEnd(function(ctx, callback) {
    var body = Buffer.concat(chunks);
    var contentType = ctx.serverToProxyResponse.headers['content-type'];
    if (contentType && contentType.indexOf('text/html') === 0) {
      body = injectJS(host, url, body) || body;
    }
    ctx.proxyToClientResponse.write(body);
    return callback();
  });
  callback();
});


function main() {
  getDomainWhiteList();
  proxy.listen({ port: port });
  console.log('listening on ' + port);
}

main();


bookmarklet.js
//
// Mobile Viewport Control v0.2.0
//
// Copyright (c) 2016 Shaun Williams
// ISC License
//
// GitHub: https://github.com/shaunstripe/mobile-viewport-control
//

//---------------------------------------------------------------------------
// JS Module Boilerplate
//---------------------------------------------------------------------------

(function (root, factory) {
  if (typeof define === 'function' && define.amd) {
    define('mobile-viewport-control', [], factory);
  }
  else if (typeof module === 'object' && module.exports) {
    module.exports = factory();
  }
  else {
    root.mobileViewportControl = factory();
  }
}(this, function() {

//---------------------------------------------------------------------------
// Getting/Setting Scroll position
//---------------------------------------------------------------------------

function getScroll() {
  return {
    left: document.body.scrollLeft,
    top: document.body.scrollTop,
  };
}

function setScroll(scroll) {
  document.body.scrollLeft = scroll.left;
  document.body.scrollTop = scroll.top;
}

//---------------------------------------------------------------------------
// Getting Initial Viewport from <meta name='viewport'> tags
// but we also include implicit defaults.
//---------------------------------------------------------------------------

function getInitialViewport() {
  // These seem to be the defaults
  var viewport = {
    'user-scalable': 'yes',
    'minimum-scale': '0',
    'maximum-scale': '10'
  };
  var tags = document.querySelectorAll('meta[name=viewport]');
  var i,j,tag,content,keyvals,keyval;
  for (i=0; i<tags.length; i++) {
    tag = tags[i];
    content = tag.getAttribute('content');
    if (tag.id !== hookID && content) {
      keyvals = content.split(',');
      for (j=0; j<keyvals.length; j++) {
        keyval = keyvals[j].split('=');
        if (keyval.length === 2) {
          viewport[keyval[0].trim()] = keyval[1].trim();
        }
      }
    }
  }
  return viewport;
}

//---------------------------------------------------------------------------
// Calculating current viewport scale
// simplified from: http://menacingcloud.com/?c=viewportScale
//---------------------------------------------------------------------------

function getOrientation() {
  var degrees = window.orientation;
  var w = document.documentElement.clientWidth;
  var h = document.documentElement.clientHeight;
  if(degrees === undefined) {
    return (w > h) ? 'landscape' : 'portrait';
  }
  return (degrees % 180 === 0) ? 'portrait' : 'landscape';
}

function getOrientedScreenWidth() {
  var orientation = getOrientation();
  var sw = screen.width;
  var sh = screen.height;
  return (orientation === 'portrait') ? Math.min(sw,sh) : Math.max(sw,sh);
}

function getScale() {
  var visualViewportWidth = window.innerWidth;
  var screenWidth = getOrientedScreenWidth();
  return screenWidth / visualViewportWidth;
}

//---------------------------------------------------------------------------
// State and Constants
//---------------------------------------------------------------------------

// A unique ID of the meta-viewport tag we must create
// to hook into and control the viewport.
var hookID = '__mobileViewportControl_hook__';

// A unique ID of the CSS style tag we must create to
// add rules for hiding the body.
var styleID = '__mobileViewPortControl_style__';

// An empirical guess for the maximum time that we have to
// wait before we are confident a viewport change has registered.
var refreshDelay = 200;

// Original viewport state before freezing.
var originalScale;
var originalScroll;

// Classes we use to make our css selector specific enough
// to hopefully override all other selectors.
// (mvc__ = mobileViewportControl prefix for uniqueness)
var hiddenClasses = [
  'mvc__a',
  'mvc__lot',
  'mvc__of',
  'mvc__classes',
  'mvc__to',
  'mvc__increase',
  'mvc__the',
  'mvc__odds',
  'mvc__of',
  'mvc__winning',
  'mvc__specificity'
];

//---------------------------------------------------------------------------
// Isolating an Element
//---------------------------------------------------------------------------

function isolatedStyle(elementID) {
  var classes = hiddenClasses.join('.');
  return [
    // We effectively clear the <html> and <body> background
    // and sizing attributes.
    'html.' + classes + ',',
    'html.' + classes + ' > body {',
    '  background: #fff;',
    '  width: auto;',
    '  min-width: inherit;',
    '  max-width: inherit;',
    '  height: auto;',
    '  min-height: inherit;',
    '  max-height: inherit;',
    '  margin: 0;',
    '  padding: 0;',
    '  border: 0;',
    '}',
    // hide everything in the body...
    'html.' + classes + ' > body > * {',
    '  display: none !important;',
    '}',
    // ...except the given element ID
    'html.' + classes + ' > body > #' + elementID + ' {',
    '  display: block !important;',
    '}'
  ].join('\n');
}

function isolate(elementID) {
  // add classes to body tag to isolate all other elements
  var classes = hiddenClasses.join(' ');
  var html = document.documentElement;
  html.className += ' ' + classes;

  // add isolating style rules
  var style = document.createElement('style');
  style.id = styleID;
  style.type = 'text/css';
  style.appendChild(document.createTextNode(isolatedStyle(elementID)));
  document.head.appendChild(style);
}

function undoIsolate() {
  // remove isolating classes from body tag
  var classes = hiddenClasses.join(' ');
  var html = document.documentElement;
  html.className = html.className.replace(classes, '');

  // remove isolating style rules
  var style = document.getElementById(styleID);
  document.head.removeChild(style);
}

//---------------------------------------------------------------------------
// Freezing and Thawing
//---------------------------------------------------------------------------

// Freeze the viewport to a given scale.
function freeze(scale) {
  // optional arguments
  var isolateID, onDone;

  // get optional arguments using their type
  var args = Array.prototype.slice.call(arguments, 1);
  if (typeof args[0] === 'string') {
    isolateID = args[0];
    args.splice(0, 1);
  }
  if (typeof args[0] === 'function') {
    onDone = args[0];
  }

  // isolate element if needed
  if (isolateID) {
    isolate(isolateID);
  }

  // freeze viewport
  var hook = document.getElementById(hookID);
  if (!hook) {
    originalScale = getScale();
    originalScroll = getScroll();
    hook = document.createElement('meta');
    hook.id = hookID;
    hook.name = 'viewport';
    document.head.appendChild(hook);
  }
  hook.setAttribute('content', [
    'user-scalable=no',
    'initial-scale='+scale,
    'minimum-scale='+scale,
    'maximum-scale='+scale
  ].join(','));

  if (onDone) {
    setTimeout(onDone, refreshDelay);
  }
}

// Thaw the viewport, restoring the scale and scroll to what it
// was before freezing.
function thaw(onDone, testEvts) {
  // restore body visibility
  var style = document.getElementById(styleID);
  if (style) {
    undoIsolate();
  }

  // exit if there is nothing to thaw
  var hook = document.getElementById(hookID);
  if (!hook) {
    return;
  }

  var initial = getInitialViewport();

  // VIEWPORT RESTORATION seems to have to be done in a certain order,
  // with some pauses in between to allow them to properly register.
  //
  // 1. Restore the user's manual zoom.
  // 2. Restore the page's zoom bounds.
  // 3. Restore the viewport size.
  // 4. Restore the viewport position (scroll).
  //
  // (Steps are listed individually below)


  // 1. Restore the user's manual zoom.
  hook.setAttribute('content', [
    'initial-scale='+originalScale,
    'minimum-scale='+originalScale,
    'maximum-scale='+originalScale,
  ].join(','));

  setTimeout(function() {
    if (testEvts && testEvts.onRestoreScale)
      testEvts.onRestoreScale();

    // 2. Restore the page's zoom bounds.
    hook.setAttribute('content', [
      'user-scalable='+initial['user-scalable'],
      'initial-scale='+originalScale,
      'minimum-scale='+initial['minimum-scale'],
      'maximum-scale='+initial['maximum-scale']
    ].join(','));

    setTimeout(function(){
      if (testEvts && testEvts.onRestoreBounds)
        testEvts.onRestoreBounds();

      // 3. Restore the viewport size.
      if (initial.width) {
        hook.setAttribute('content', [
          'width='+initial['width']
        ].join(','));
      }

      setTimeout(function() {
        // Remove our meta viewport hook.
        document.head.removeChild(hook);

        // 4. Restore the viewport position (scroll).
        setScroll(originalScroll);

        if (testEvts && testEvts.onRestoreScroll)
          testEvts.onRestoreScroll();

        if (onDone)
          onDone();

      }, refreshDelay);
    }, refreshDelay);
  }, refreshDelay);
}

//---------------------------------------------------------------------------
// Public API
//---------------------------------------------------------------------------

return {
  getInitialViewport: getInitialViewport,
  getScale: getScale,
  isolate: isolate,
  undoIsolate: undoIsolate,

  // stable
  version: '0.2.0',
  freeze: freeze,
  thaw: thaw
};

})); // end module scope
(function(){
  var viewport = window.mobileViewportControl;
  if (viewport) {
    test();
  } else {
    require(['mobile-viewport-control'], function(module) {
      viewport = module;
      test();
    });
  }

  function test() {
    var isolated = createIsolatedElement();
    var initial = JSON.stringify(viewport.getInitialViewport());
    alert("About to freeze viewport on a custom element... Viewport is: "+initial);
    viewport.freeze(1, isolated.id);
  }

  function createIsolatedElement() {
    var isolated = document.createElement('div');
    isolated.style.cssText = 'color: #000; text-align: left; font-size: 12px; padding: 20px;';
    isolated.id = 'myIsolatedElement';
    isolated.innerHTML = 'I should be the only visible thing, and you should not be able to pinch zoom or scroll.<br>';

    var btn = document.createElement('button');
    btn.innerHTML = 'Click me to restore!';
    btn.onclick = function() {
      document.body.removeChild(isolated);
      viewport.thaw();
    }
    isolated.appendChild(btn);

    document.body.appendChild(isolated);
    return isolated;
  }
}());
